draft-ietf-tls-esni.md

```diff
@@ -1012,39 +985,60 @@ server to include any information it requires to process the second ClientHello.
+The backend server embeds in ServerHello.random a string derived from the inner
+handshake. It begins by computing its ServerHello as usual, except the last 8
+bytes of ServerHello.random are set to zero. It then computes the transcript
+hash for ClientHelloInner up to and including the modified ServerHello, as
+described in {{RFC8446}}, Section 4.4.1. Let transcript_ech_conf denote the
+output. Finally, the backend server overwrites the last 8 bytes of the
+ServerHello.random with the following string:
 
 ~~~
-   accept_confirmation = HKDF-Expand-Label(0,
+   accept_confirmation = HKDF-Expand-Label(
+      HKDF-Extract(0, ClientHelloInner.random),
       "ech accept confirmation",
-      Transcript-Hash(ClientHelloInner...ServerHelloECHConf),
+      transcript_ech_conf,
       8)
 ~~~
 
-where HKDF-Expand-Label and Transcript-Hash are as defined in {{RFC8446}},
-Section 7.1, "0" indicates a string of Hash.length bytes set to zero, and
-ClientHelloInner...ServerHelloECHConf refers to the sequence of handshake
-messages beginning with the first ClientHelloInner and ending with
-ServerHelloECHConf. (Note that ClientHelloInner and ServerHelloECHConf
-messages replace the ClientHello and ServerHello messages in the transcript
-hash sequence as specified in Section 4.1.1 of {{RFC8446}}. Finally, the
-backend server constructs its ServerHello message so that it is equal to
-ServerHelloECHConf but with the last 8 bytes of ServerHello.random set to
-`accept_confirmation`.
+where HKDF-Expand-Label is defined in {{RFC8446}}, Section 7.1, "0" indicates a
+string of Hash.length bytes set to zero, and Hash is the hash function used to
+compute the transcript hash.
 
 The backend server MUST NOT perform this operation if it negotiated TLS 1.2 or
 below. Note that doing so would overwrite the downgrade signal for TLS 1.3 (see
 {{RFC8446}}, Section 4.1.3).
 
-The "ech_is_inner" is expected to have an empty payload. If the payload is
-non-empty (i.e., the length of the "extension_data" field is non-zero) then the
+### Sending HelloRetryRequest {#backend-server-hrr}
+
+When the backend server sends HelloRetryRequest in response to the ClientHello,
+it similarly confirms ECH acceptance by adding a confirmation signal to its
+HelloRetryRequest. But instead of embedding the signal in the
+HelloRetryRequest.random (the value of which is specified by {{RFC8446}}), it
+sends the signal in an extension.
+
+The backend server begins by computing HelloRetryRequest as usual, except that
+it also contains an "encrypted_client_hello" extension with a payload of 8 zero
+bytes. It then computes the transcript has for ClientHelloInner1 up to and
+including the modified HelloRetryRequest. Let transcript_hrr_ech_conf denote the
+output. Finally, the backend server overwrites the payload of the
+"encrypted_client_hello" extension with the following string:
+
+~~~
+   accept_confirmation = HKDF-Expand-Label(
+      HKDF-Extract(0, ClientHelloInner1.random),
```

Comment:
If you drop `HKDF-Extract(0, ClientHelloInner.random)` above, presumably you should drop `HKDF-Extract(0, ClientHelloInner1.random)` too.